sympy/tensor/array/arrayop.py

```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
 
     """
     from sympy.matrices import MatrixBase
+    from sympy.tensor import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
-
     if isinstance(dx, array_types):
         dx = ImmutableDenseNDimArray(dx)
         for i in dx:
             if not i._diff_wrt:
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
+        if isinstance(expr, NDimArray):
+            expr = expr.as_immutable()
+        else:
+            expr = ImmutableDenseNDimArray(expr)
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
```

Comment:
if both expr and dx are sparse, just use a dict comprehension (in a single line).

Reply:
The line 213 uses dict comprehension when expr is sparse but dx is not. 
I think for the case where `dx` is an array(sparse or dense), it is necessary to calculate the new index for the derivative by each term in `dx` , which is the purpose of `_sparse_array_derive_by_array`.

Reply:
Yes, so why not just an array comprehension?

Reply:
`{k1+k2: v2.diff(v1) for k1, v1 in dx... dict ...items() for k2, v2 in expr... dict ...items()}`

Reply:
OK, I got your point! Thank you!